### PR TITLE
[MCPG-15] docs: add agent gateway

### PR DESCRIPTION
**Pull request by polarweasel**
Adding the first pass at the Agent Gateway docs, along with some updated agent instructions for more-automated docs production.

## Validation

List the commands you ran.

```bash
fern check
fern docs md check
rumdl check docs
```

## Checklist

- [x] I am familiar with the [Contributing Guidelines](https://github.com/NVIDIA/dsx-exchange/blob/main/CONTRIBUTING.md).
- [x] Documentation updated, if needed.
- [ ] Tests or validation added/updated, if needed.
- [x] License headers are present on applicable source files.
- [ ] Third-party license inventory updated, if dependencies changed.
- [ ] Security-sensitive changes were reviewed privately before public disclosure.
- [x] My commits are signed off (`git commit -s`) per the DCO.

## Summary by CodeRabbit

* **Documentation**
  * Added comprehensive Agent Gateway documentation covering architecture, deployment, configuration, operations, MCP client usage, publishing, and supported capabilities.
  * Added writing and documentation guidelines for consistent, verified content.
  * Updated architecture, navigation, and Getting Started documentation with Agent Gateway, Event Bus, and QoS information.
  * Added an Agent Gateway overview link and improved Markdown formatting and numbering consistency across existing guides.
* **Chores**
  * Updated documentation linting rules and repository documentation configuration.

### CODE REVIEW

File: docs/architecture.md [hunk 1]
```diff
@@ -1,21 +1,30 @@
...
 
-DSX Exchange consists of three components:
+DSX Exchange consists of the following components:
 
 | Component | What It Is |
 |-----------|------------|
 | DSX Event Bus | NATS with MQTT 3.1.1, HA clustering, and leaf-node federation |
 | AsyncAPI Schema | Formal topic definitions and payload contracts per service team |
 | Auth-Callout Service | OAuth2/mTLS/NKey authentication with topic-level ACLs |
+| DSX Agent Gateway | Authenticated Model Context Protocol routing to local and remote MCP servers |
```
Comment (FrankSpitulski) [inline docs/architecture.md:16]: missing mention of the bridge, which is the main extension we provide over upstream agent gateway.

Comment (FrankSpitulski) [reply docs/architecture.md:16]: since auth-callout is mentioned, it would be consistent to call out the bridge

**Review by FrankSpitulski — DISMISSED** [review]

File: docs/agent-gateway/architecture.mdx [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+
+Direct and bridged requests have different state boundaries.
+
+For direct routing, a client can initialize a streamable HTTP MCP session and reuse its `Mcp-Session-Id` value. Selector-mode routing keeps that session associated with one discovered upstream endpoint. The tested dataplane preserves this upstream association when a later request reaches another gateway replica.
+
+The bridge does not extend the caller's MCP session into a remote shard. The hub and leaves remove `Mcp-Session-Id` before the leaf calls its local gateway. Remote requests are independent, stateless MCP requests.
+
+The hub handles `initialize` and `ping` locally. It also accepts supported JSON-RPC notifications without forwarding session state. Methods that require server-side MCP session state, client state, or resource-path routing are not available through the bridge.
+
+A bridge leaf temporarily owns the response state for an active SSE stream. The hub reads ordered frames from that leaf until completion, cancellation, or idle expiry. This temporary stream ownership does not create an MCP session in the remote shard.
```
Comment (FrankSpitulski) [inline docs/agent-gateway/architecture.mdx]: maybe too much impl detail. this isn't arch.

File: docs/agent-gateway/architecture.mdx [hunk 3]
```diff
@@ -0,0 +1,126 @@
...
+
+The local gateway repeats JWT verification, tenant derivation, authorization, and rate limiting for the forwarded request. The bridge preserves the caller's `Authorization` header across this path.
+
+The bridge supports JSON responses and streamed server-sent events (SSE) for routed invocations. The [validated capabilities](validated-capabilities.mdx) page identifies methods that the stateless bridge does not support.
+
+## Sessions and State
+
+Direct and bridged requests have different state boundaries.
+
+For direct routing, a client can initialize a streamable HTTP MCP session and reuse its `Mcp-Session-Id` value. Selector-mode routing keeps that session associated with one discovered upstream endpoint. The tested dataplane preserves this upstream association when a later request reaches another gateway replica.
```
Comment (FrankSpitulski) [inline docs/agent-gateway/architecture.mdx]: don't need the tested or validated qualifier. the features we include are tested. omitted tests are a bug, never on purpose.

File: docs/agent-gateway/client-quickstart.mdx [hunk 4]
```diff
@@ -0,0 +1,174 @@
...
+
+## Prerequisites
+
+Obtain the following values from the gateway operator.
+
+- The complete HTTPS gateway endpoint, including `/mcp`.
+- A bearer token from a JSON Web Token (JWT) provider configured for the gateway.
+- Access to at least one MCP target.
+
+Install `curl` 8.3.0 or later, `jq`, and `awk` before you run the examples. The examples import the token into `curl` without expanding it in the shell command.
```
Comment (FrankSpitulski) [inline docs/agent-gateway/client-quickstart.mdx]: manual curl may be a bit much in terms of ease of use. could we include a sample with some popular cli instead? or with an agent?

File: docs/agent-gateway/configuration-reference.mdx [hunk 5]
```diff
@@ -0,0 +1,164 @@
+---
+title: Configuration Reference
```
Comment (FrankSpitulski) [inline docs/agent-gateway/configuration-reference.mdx:2]: strong chance for drift here. can we link the values.yaml or helm readme instead?

File: docs/agent-gateway/deployment.mdx [hunk 6]
```diff
@@ -0,0 +1,262 @@
+---
+title: Deploy DSX Agent Gateway
+sidebar-title: Deployment
```
Comment (FrankSpitulski) [inline docs/agent-gateway/deployment.mdx:3]: again, this looks like it has a lot of duplication with the chart readme

File: docs/agent-gateway/operations.mdx [hunk 7]
```diff
@@ -0,0 +1,278 @@
...
+
+```bash
+kubectl describe gateway dsx-agent-gateway --namespace dsx-agent-gateway
+kubectl describe httproute dsx-agent-gateway-mcp-route \
+  --namespace dsx-agent-gateway
+```
+
+## Check Workload Health
+
+Kubernetes readiness is the primary health signal for chart-managed workloads. The rate-limit service uses `/healthcheck` for readiness and liveness probes. Both bridge roles use `/readyz` and `/livez` on port `3001`. Hub readiness requires a NATS connection and a completed shard-discovery cache. Leaf readiness requires a NATS connection. List Pods that are not ready:
```
Comment (FrankSpitulski) [inline docs/agent-gateway/operations.mdx]: these sorts of details about ports and paths aren't necessary. the charts show them, and they're  not needed to view pod status. they're already wired up in k8s.

File: docs/agent-gateway/operations.mdx [hunk 8]
```diff
@@ -0,0 +1,278 @@
...
+
+## Configure Metrics
+
+Metrics require Prometheus Operator CRDs and a scraper that discovers monitor resources in the release namespace. The chart creates `PodMonitor` or `ServiceMonitor` resources for these components when metrics are enabled:
+
+- Agentgateway dataplane.
+- Bundled Agentgateway controller.
+- Rate-limit service.
+- Enabled bridge.
+The upstream Valkey chart controls its own exporter and `ServiceMonitor`. Configure the shared chart metrics and Valkey metrics together:
```
Comment (FrankSpitulski) [inline docs/agent-gateway/operations.mdx]: readme contains observability config and chart samples. maybe this doc should include usage demonstrations and images of the deployed dashboards?

File: docs/agent-gateway/overview.mdx [hunk 9]
```diff
@@ -0,0 +1,67 @@
+---
+title: DSX Agent Gateway Overview
+sidebar-title: Overview
+position: 10
+last-updated: August 5, 2026
+---
+
+DSX Agent Gateway provides authenticated, tenant-aware ingress for Model Context Protocol (MCP) traffic. It gives MCP clients one `/mcp` endpoint for discovering and invoking approved MCP servers.
```
Comment (FrankSpitulski) [inline docs/agent-gateway/overview.mdx]: dup intro with architecture

File: docs/agent-gateway/overview.mdx [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+
+| Audience | Primary Tasks |
+| -------- | ------------- |
+| MCP client developer | Connect a client, initialize MCP, discover capabilities, and invoke an approved tool or prompt. |
+| MCP server developer | Publish a server through Kubernetes discovery or a static upstream definition. |
+| Platform operator | Deploy the chart, configure identity and tenant policy, connect an optional bridge, and operate the service. |
+
+## How Requests Reach MCP Servers
+
+For a direct request, the gateway verifies the caller's JWT and derives a tenant ID from verified claims. It applies tenant authorization and rate limiting before it routes the request to an allowed MCP target.
```
Comment (FrankSpitulski) [inline docs/agent-gateway/overview.mdx]: what would be an indirect request?

File: docs/agent-gateway/overview.mdx [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+
+The gateway supports two target types:
+
+- A selector target discovers Kubernetes Service ports marked with `appProtocol: agentgateway.dev/mcp`.
+- A static target connects to a configured HTTP or HTTPS MCP endpoint.
+
+The optional bridge adds cross-shard routing. An entry gateway routes to a bridge hub, and bridge leaves forward requests through the gateway in each selected shard. The bridge uses NATS request-reply traffic provided by the DSX Event Bus.
```
Comment (FrankSpitulski) [inline docs/agent-gateway/overview.mdx]: shard needs definition

File: docs/agent-gateway/publish-mcp-server.mdx [hunk 12]
```diff
@@ -0,0 +1,197 @@
...
+
+Your MCP server remains responsible for authorizing the caller within its domain. Do not use caller-supplied tenant headers as an authenticated identity.
+
+Use the [architecture guide](architecture.mdx) for the complete direct and bridged request paths.
+
+## Prepare the MCP Server
+
+The server must meet the following requirements.
+
+- Serve MCP through the Streamable HTTP transport at `/mcp`.
```
Comment (FrankSpitulski) [inline docs/agent-gateway/publish-mcp-server.mdx]: this is a very long way of saying supports the MCP protocol

File: docs/agent-gateway/publish-mcp-server.mdx [hunk 13]
```diff
@@ -0,0 +1,197 @@
...
+  ports:
+    - name: mcp
+      port: 3001
+      targetPort: mcp
+      appProtocol: agentgateway.dev/mcp
+```
+
+The Service selector must identify the Pods that serve `/mcp` on the named target port. Configure startup, readiness, and liveness probes on the workload before you publish it.
+
+## Configure Service Discovery
```
Comment (FrankSpitulski) [inline docs/agent-gateway/publish-mcp-server.mdx:61]: this is generally not required unless you're using the namespace restrictor feature

Comment (FrankSpitulski) [reply docs/agent-gateway/publish-mcp-server.mdx:61]: nvm wrong section

File: docs/agent-gateway/publish-mcp-server.mdx [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+## Understand Published Names
+
+When the gateway aggregates multiple targets, it prefixes tools and prompts with their target name. The published name has the form `<target>_<name>`.
+
+For example, the `lookup` tool from `inventory-mcp` is published as `inventory-mcp_lookup`. Clients must use the complete published name in `tools/call`. An unprefixed tool or prompt name does not select a default target.
+
+Choose stable, lowercase, hyphenated Service, port, and static target names. Changing a target name also changes its published tool and prompt names.
+
+## Apply the Configuration
```
Comment (FrankSpitulski) [inline docs/agent-gateway/publish-mcp-server.mdx:141]: rather link to deployment

File: docs/agent-gateway/validated-capabilities.mdx [hunk 15]
```diff
@@ -0,0 +1,92 @@
...
+last-updated: August 5, 2026
+---
+
+This page records DSX Agent Gateway behavior exercised by checked-in tests. It is not a complete Model Context Protocol conformance statement.
+
+## MCP Method Coverage
+
+The direct path sends a request from the entry gateway to a local or static MCP target. The bridge path sends a request through a hub, the DSX Event Bus, and a leaf to a shard gateway.
+
+| Capability | Direct path | Bridge path |
```
Comment (FrankSpitulski) [inline docs/agent-gateway/validated-capabilities.mdx]: no need to specify validation.

File: docs/agent-gateway/operations.mdx (hunk 8 above)
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/agent-gateway/operations.mdx]: Fixed in an upcoming commit

File: docs/AGENTS.md [hunk 16]
```diff
@@ -0,0 +1,145 @@
...
+documentation. Treat `docs/` as the source of truth for published content.
+
+## Role
+
+- Write clear, accurate, task-oriented documentation for developers and
+  operators who deploy, integrate with, and operate DSX Exchange.
+- Distinguish the DSX Event Bus, AsyncAPI schemas, and DSX Agent Gateway while
+  explaining how they work together as DSX Exchange.
+- Preserve the reader's workflow: explain what to do, when to do it, and how
+  to verify it.
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/AGENTS.md]: I've updated the rule to be one paragraph per line, and reflowed the text in this file. It'll be committed shortly.

File: docs/agent-gateway/architecture.mdx [hunk 17]
```diff
@@ -0,0 +1,126 @@
...
+flowchart LR
+    client["MCP Client"] -->|"HTTPS /mcp"| edge["Operator-Owned Edge"]
+    edge -->|"HTTP /mcp"| gateway["Agentgateway Dataplane"]
+    gateway --> jwt["Strict JWT Verification"]
+    jwt --> tenant["Tenant ID From Verified Claims"]
+    tenant --> limit["Tenant Rate Limit"]
+    limit <-->|"Counters"| valkey["Valkey"]
+    limit --> authz["MCP Target Authorization"]
+    authz --> route["Target Selection"]
+    route -->|"Original Bearer Token"| service["MCP Server"]
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/agent-gateway/architecture.mdx:42]: @FrankSpitulski does this one make sense to you?

File: docs/architecture.md [hunk 18]
```diff
@@ -1,21 +1,30 @@
+---
+last-updated: August 5, 2026
+---
+
 # Architecture
 
 ## DSX Exchange
 
 DSX Exchange is an open source event bus for AI factory operations. It supports real-time operational signal exchange between power management systems, building management systems, cooling infrastructure, grid interfaces, and compute schedulers.
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/architecture.md]: @FrankSpitulski how's this one grab you?

File: docs/architecture.md [hunk 19]
```diff
@@ -1,21 +1,30 @@
...
 
 | Component | What It Is |
 |-----------|------------|
 | DSX Event Bus | NATS with MQTT 3.1.1, HA clustering, and leaf-node federation |
 | AsyncAPI Schema | Formal topic definitions and payload contracts per service team |
 | Auth-Callout Service | OAuth2/mTLS/NKey authentication with topic-level ACLs |
+| DSX Agent Gateway | Authenticated Model Context Protocol routing to local and remote MCP servers |
 
 <Info title="GitHub Repository">
   The DSX Exchange GitHub repository is available at [https://github.com/NVIDIA/dsx-exchange](https://github.com/NVIDIA/dsx-exchange).
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/architecture.md]: fixed in an upcoming commit

File: docs/AGENTS.md [hunk 20]
```diff
@@ -0,0 +1,145 @@
+# Documentation Agent Guide
+
+You are a documentation engineer and writer for DSX Exchange user-facing
+documentation. Treat `docs/` as the source of truth for published content.
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/AGENTS.md]: I've adjusted the instructions in docs/AGENTS.md to reflect a one-paragraph-per-line approach. It'll get committed soon.

File: docs/architecture.md (hunk 1 above)
Comment (polarweasel) [reply docs/architecture.md:16]: Added

File: docs/agent-gateway/architecture.mdx (hunk 2 above)
Comment (polarweasel) [reply docs/agent-gateway/architecture.mdx]: adjusted...

File: docs/agent-gateway/client-quickstart.mdx (hunk 4 above)
Comment (polarweasel) [reply docs/agent-gateway/client-quickstart.mdx]: This is now a tabbed interface: the default tab walks you through a tool call with MCP Inspector, and the second tab is the curl calls.

File: docs/agent-gateway/configuration-reference.mdx (hunk 5 above)
Comment (polarweasel) [reply docs/agent-gateway/configuration-reference.mdx:2]: Reduced the detail level in this one, and linked out to the values file

File: docs/agent-gateway/deployment.mdx (hunk 6 above)
Comment (polarweasel) [reply docs/agent-gateway/deployment.mdx:3]: Reduced detail level

File: docs/agent-gateway/operations.mdx (hunk 7 above)
Comment (polarweasel) [reply docs/agent-gateway/operations.mdx]: Detail reduction implemented here too

File: docs/agent-gateway/operations.mdx (hunk 8 above)
Comment (polarweasel) [reply docs/agent-gateway/operations.mdx]: Any suggestions for usage demos, or sources for images you'd like to see?

File: docs/agent-gateway/overview.mdx (hunk 9 above)
Comment (polarweasel) [reply docs/agent-gateway/overview.mdx]: fixed

File: docs/agent-gateway/overview.mdx (hunk 10 above)
Comment (polarweasel) [reply docs/agent-gateway/overview.mdx]: Clarified what happens with and without the bridge.

File: docs/agent-gateway/overview.mdx (hunk 11 above)
Comment (polarweasel) [reply docs/agent-gateway/overview.mdx]: Added definition on first use earlier in the file

File: docs/agent-gateway/publish-mcp-server.mdx (hunk 12 above)
Comment (polarweasel) [reply docs/agent-gateway/publish-mcp-server.mdx]: Fixed 😁

File: docs/agent-gateway/publish-mcp-server.mdx (hunk 14 above)
Comment (polarweasel) [reply docs/agent-gateway/publish-mcp-server.mdx:141]: Fixed

File: docs/agent-gateway/validated-capabilities.mdx (hunk 15 above)
Comment (polarweasel) [reply docs/agent-gateway/validated-capabilities.mdx]: Lots of adjustments made on this one (including renaming to supported capabilities)

File: docs/agent-gateway/configuration-reference.mdx [hunk 21]
```diff
@@ -0,0 +1,164 @@
...
+`auth.jwt.providers` must contain at least one provider. Each map key must be a lowercase DNS label with no more than 56 characters.
+
+Each provider requires the following settings.
+
+| Setting | Requirement |
+| ------- | ----------- |
+| `issuer` | Must be unique across configured providers. |
+| `audiences` | Must contain at least one accepted token audience. |
+| `jwksUrl` | Must be a complete HTTP or HTTPS URL with a host and path. |
+| `tenantIdExpression` | Must derive a nonempty tenant ID from verified `jwt.*` claims. |
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/agent-gateway/configuration-reference.mdx]: This is beyond the scope of documentation. @FrankSpitulski do you concur?

File: docs/agent-gateway/configuration-reference.mdx [hunk 22]
```diff
@@ -0,0 +1,164 @@
...
+```yaml
+upstreams:
+  vendor-mcp:
+    mode: static
+    address: https://mcp.vendor.example/mcp
+    protocol: StreamableHTTP
+    requestTimeout: 10s
+```
+
+`protocol` accepts `StreamableHTTP` or `SSE` for a static upstream. It defaults to `StreamableHTTP`. An HTTPS address enables Transport Layer Security between the gateway and that upstream.
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/agent-gateway/configuration-reference.mdx]: Again, beyond the scope of the docs, right @FrankSpitulski?

File: docs/agent-gateway/operations.mdx [hunk 23]
```diff
@@ -0,0 +1,278 @@
...
+## Scale the Workloads
+
+Change replica counts through Helm values so later releases preserve the intended scale. The chart uses these defaults:
+
+| Component | Default Scale | Value |
+| --------- | ------------- | ----- |
+| Agentgateway dataplane | 3 replicas | `gateway.replicaCount` |
+| Rate-limit service | 2 replicas | `rateLimit.replicaCount` |
+| Enabled bridge | 2 replicas | `bridge.replicaCount` |
+| Bundled Agentgateway controller | 1 replica | Native Agentgateway chart values |
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/agent-gateway/operations.mdx]: Adjusted, following the pattern of full name on first use in a file, followed by the shorter "Agent Gateway" on subsequent use.

**Comment by polarweasel** [comment]
@coderabbitai another review, please!

File: docs/agent-gateway/architecture.mdx [hunk 24]
```diff
@@ -0,0 +1,126 @@
+---
+title: DSX Agent Gateway Architecture
+sidebar-title: Architecture
+position: 20
+last-updated: August 5, 2026
+---
+
+DSX Agent Gateway provides a policy enforcement and routing layer between MCP clients and MCP servers. Without bridge routing, the gateway sends requests to targets that it can reach. The optional bridge extends selected requests to remote gateway shards through the DSX Event Bus.
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply docs/agent-gateway/architecture.mdx]: MCP, JWT, SSE, and also DSX are all common acronyms in these docs. No need to spell out on first use.

File: WRITING.md [hunk 25]
```diff
@@ -0,0 +1,120 @@
...
+| Test title | `works after retry` | `reconnects the CPC leaf node after the CSC listener restarts` |
+| Test title | `covers edge cases` | `rejects overlapping cross-layer subject patterns` |
+| Test title | `does the right thing for auth` | `denies an MQTT publish outside the account subject allowlist` |
+| Pull request discussion | `This seems brittle.` | `This lookup assumes every Gateway has an assigned address. Pending Gateways cause an empty endpoint.` |
+| Pull request discussion | `Can we clean this up?` | `These templates build the same listener name. Define the name in one Helm helper.` |
+| Pull request discussion | `Make this more robust.` | `Return an error when the bridge cannot resolve the upstream service, and add a resolution-failure test.` |
+| Pull request discussion | `This is a small change.` | `This change updates one Helm value and does not change the rendered Service ports.` |
+| Announcement | `Improved local deployment.` | `Local deployment now reuses unchanged Agent Gateway images.` |
+| Release entry | `Fixed various issues.` | `The Agent Gateway chart now rejects a listener with no protocol.` |
```
[the remark this answers is not in the document]
Comment (polarweasel) [reply WRITING.md:118]: My approach in these docs is to use the full product name on first mention on every page, after which we can shorten to Agent Gateway.

**Review by shassard-nvidia — APPROVED** [review]
lgtm
